### PR TITLE
[IMP] mail: initialize use readonly route

### PR DESCRIPTION
**Pull request by phenix-factory**
\* = im_livechat, mail_bot, test_discuss_full

OdooBot onboarding needs to be moved to webclient load.

Write access is not useful since settings were moved out of the init messaging.

See https://github.com/odoo/odoo/pull/150625

https://github.com/odoo/enterprise/pull/69565

### CODE REVIEW

File: addons/web/tests/test_session_info.py [hunk 1]
```diff
@@ -38,6 +38,8 @@ def setUpClass(cls):
     def test_session_info(self):
         """ Checks that the session_info['user_companies'] structure correspond to what is expected """
         self.authenticate(self.user.login, self.user_password)
+        # avoid the create part of res.users.settings since this route is readonly
+        self.env['res.users.settings']._find_or_create_for_user(self.user)
         response = self.url_open("/web/session/get_session_info", data=self.payload, headers=self.headers)
```
Comment (alexkuhn) [inline addons/web/tests/test_session_info.py:43]: Nitpicking: the comment refers to the `url_open` comment, so it's a bit confusing to read so it takes more time than necessary. Probably saying `get_session_info is readonly so let's create beforehand` would be less effort to understand these LOCs and the comment.

**Review by alexkuhn — COMMENTED** [review]
LGTM

File: addons/mail_bot/models/res_users.py [hunk 2]
```diff
@@ -25,12 +26,17 @@ class ResUsers(models.Model):
     def SELF_READABLE_FIELDS(self):
         return super().SELF_READABLE_FIELDS + ['odoobot_state']
 
-    def _init_messaging(self, store):
+    @api.model
+    def _on_webclient_bootstrap(self):
+        if self.odoobot_state in [False, 'not_initialized'] and self.env.user and self.env.user._is_internal():
+            self.env.user._init_odoobot()
```
Comment (Julien00859) [inline addons/mail_bot/models/res_users.py:32]: `@api.model` is a decorator for rpc, it makes that the method can be called without a list of ids (i.e. with an empty recordset). Since this method is private, it cannot be called via rpc, the decorator is useless.

You call this method with an empty recordset, `self.odoobot_state` is always `False`

File: addons/mail_bot/models/res_users.py [hunk 3]
```diff
@@ -25,12 +26,17 @@ class ResUsers(models.Model):
...
+
+    @api.model
+    def _init_store_data(self, store: Store, /):
         odoobot_onboarding = False
-        if self.odoobot_state in [False, 'not_initialized'] and self._is_internal():
+        if self.odoobot_state in [False, 'not_initialized'] and self.env.user and self.env.user._is_internal():
```
Comment (Julien00859) [inline addons/mail_bot/models/res_users.py:37]: same

**Review by Julien00859 — CHANGES_REQUESTED** [review]

File: addons/web/controllers/home.py [hunk 4]
```diff
@@ -61,6 +61,7 @@ def web_client(self, s_action=None, **kw):
         # Restore the user on the environment, it was lost due to auth="none"
         request.update_env(user=request.session.uid)
         try:
+            request.env["res.users"]._on_webclient_bootstrap()
```
Comment (Julien00859) [inline addons/web/controllers/home.py:64]: inside `web` we care about the quotes enough to bother you about them :D
```suggestion
            request.env['res.users']._on_webclient_bootstrap()
```

File: addons/mail_bot/models/res_users.py (hunk 2 above)
Comment (Julien00859) [reply addons/mail_bot/models/res_users.py:32]: If this is intended, I would rather read `if not self or ...`

File: addons/mail/static/tests/discuss_app/discuss.test.js [hunk 5]
```diff
@@ -1093,7 +1093,8 @@ test("out-of-focus notif on needaction message in channel", async () => {
             }
         },
     });
-    onRpcBefore("/mail/action", async (args) => {
+    mockService("mail.out_of_focus", { setInterval: () => {} }); // so that no setInterval runs
```
Comment (seb-odoo) [inline addons/mail/static/tests/discuss_app/discuss.test.js:1096]: Why do we need to do that?

File: addons/mail_bot/models/res_users.py [hunk 6]
```diff
@@ -25,12 +26,17 @@ class ResUsers(models.Model):
...
         return super().SELF_READABLE_FIELDS + ['odoobot_state']
 
-    def _init_messaging(self, store):
+    @api.model
+    def _on_webclient_bootstrap(self):
+        if self.odoobot_state in [False, 'not_initialized'] and self.env.user and self.env.user._is_internal():
+            self.env.user._init_odoobot()
+
+    @api.model
+    def _init_store_data(self, store: Store, /):
```
Comment (seb-odoo) [inline addons/mail_bot/models/res_users.py:35]: And both methods should actually be made to work with `self` user rather than using the user from env internally.

Caller should just use `self.env.user._on_webclient_bootstrap()` as appropriate.

File: addons/mail_bot/models/res_users.py [hunk 7]
```diff
@@ -25,12 +26,17 @@ class ResUsers(models.Model):
...
         odoobot_onboarding = False
-        if self.odoobot_state in [False, 'not_initialized'] and self._is_internal():
+        if self.odoobot_state in [False, 'not_initialized'] and self.env.user and self.env.user._is_internal():
             odoobot_onboarding = True
-            self._init_odoobot()
-        super()._init_messaging(store)
+        super()._init_store_data(store)
```
Comment (seb-odoo) [inline addons/mail_bot/models/res_users.py:39]: A variable for `odoobot_onboarding` is probably not necessary anymore, the value can be directly assigned.

File: addons/web/controllers/home.py (hunk 4 above)
Comment (tde-banana-odoo) [reply addons/web/controllers/home.py:64]: But Julien, sthap !

File: addons/mail/static/tests/discuss_app/discuss.test.js (hunk 5 above)
Comment (alexkuhn) [reply addons/mail/static/tests/discuss_app/discuss.test.js:1096]: Pretty sure this was an issue when the out-of-focus code had blinking effect, which was removed recently, thus the `mockService({ setInterval })` no longer makes sense

File: addons/mail_bot/models/res_users.py (hunk 2 above)
Comment (phenix-factory) [reply addons/mail_bot/models/res_users.py:32]: I remove the useless decorator.

> If this is intended, I would rather read if not self or ...

I don't understand this part, could you please elaborate?

File: addons/web/controllers/home.py (hunk 4 above)
Comment (phenix-factory) [reply addons/web/controllers/home.py]: We are using `"` in discuss/mail and will also bother you. I guess it's fair.

File: addons/mail_bot/models/res_users.py (hunk 2 above)
Comment (Julien00859) [reply addons/mail_bot/models/res_users.py]: It is a common (arguably *bad*) practice to write:

```
if record.field == False```
to check *at the same time* that either the record is empty, either that the record exists but as field's value `False`.

In your code, since `self` is an empty recordset (called as `request.env['res.users']._on_webclient_bootstrap()`, i.e. without records) I *think* that by `if self.odoobot_state in [False, 'not_initialized']` you actually mean `if not self or self.odoobot_state == 'not_initialized'`. If it is the case (i.e. `False` isn't a valid value for the `odoobot_field`) then I would prefer we write `if not self or ...` than `if not self.odoobot_state in [False, ...]`.

File: addons/mail_bot/models/res_users.py [hunk 8]
```diff
@@ -25,13 +26,14 @@ class ResUsers(models.Model):
     def SELF_READABLE_FIELDS(self):
         return super().SELF_READABLE_FIELDS + ['odoobot_state']
 
-    def _init_messaging(self, store):
-        odoobot_onboarding = False
-        if self.odoobot_state in [False, 'not_initialized'] and self._is_internal():
-            odoobot_onboarding = True
+    def _on_webclient_bootstrap(self):
+        if self and self._is_internal() and self.odoobot_state in [False, 'not_initialized']:
```
Comment (seb-odoo) [inline addons/mail_bot/models/res_users.py:30]: Checking self is not necessary now that it is called on the actual user record.

File: addons/web/models/res_users.py [hunk 9]
```diff
@@ -21,3 +21,7 @@ def name_search(self, name='', args=None, operator='ilike', limit=100):
             if user_tuple := super().name_search(name, expression.AND([args or [], [('id', '=', uid)]]), operator, limit=1):
                 user_list = [user_tuple[0], *user_list[:-1]]
         return user_list
+
+    @api.model
+    def _on_webclient_bootstrap(self):
+        pass
```
Comment (seb-odoo) [inline addons/web/models/res_users.py:27]: Not api.model, but ensure_one could be appropriate.

File: addons/mail_bot/models/res_users.py [hunk 10]
```diff
@@ -25,13 +26,14 @@ class ResUsers(models.Model):
...
             self._init_odoobot()
-        super()._init_messaging(store)
-        store.add({"odoobotOnboarding": odoobot_onboarding})
+
+    def _init_store_data(self, store: Store, /):
```
Comment (seb-odoo) [inline addons/mail_bot/models/res_users.py:33]: `_init_store_data` should eventually be changed to also use "self" rather than env user, but this can be another PR.

File: addons/mail_bot/models/res_users.py (hunk 2 above)
Comment (phenix-factory) [reply addons/mail_bot/models/res_users.py]: I see now, thank you.
I have updated the code accordingly.

Comment (Julien00859) [reply addons/mail_bot/models/res_users.py]: Or (imo better) we could make it required, with `not_initialized` be the default value

File: addons/web/models/res_users.py [hunk 11]
```diff
@@ -21,3 +21,7 @@ def name_search(self, name='', args=None, operator='ilike', limit=100):
             if user_tuple := super().name_search(name, expression.AND([args or [], [('id', '=', uid)]]), operator, limit=1):
                 user_list = [user_tuple[0], *user_list[:-1]]
         return user_list
+
+    @api.model
```
Comment (seb-odoo) [inline addons/web/models/res_users.py:25]: `api.model` is still not necessary here (and actually false/misleading now)

**Review by Julien00859 — APPROVED** [review]
python code, but most of all to reset my "blocking" review

File: addons/mail_bot/models/res_users.py [hunk 12]
```diff
@@ -25,13 +26,16 @@ class ResUsers(models.Model):
...
+        super()._on_webclient_bootstrap()
+        if self._is_internal() and not self.odoobot_state or self.odoobot_state == "not_initialized":
             self._init_odoobot()
-        super()._init_messaging(store)
-        store.add({"odoobotOnboarding": odoobot_onboarding})
+
+    def _init_store_data(self, store: Store, /):
+        store.add({"odoobotOnboarding": False})
+        if self._is_internal() and not self.odoobot_state or self.odoobot_state == "not_initialized":
+            store.add({"odoobotOnboarding": True})
```
Comment (seb-odoo) [inline addons/mail_bot/models/res_users.py:37]: This condition is actually no longer correct. It used to be called before `_init_odoobot` so assumed it set True when it became initialized, but this way it is the inverse of what we want.

`odoobotOnboarding` should be True in any of the intermediate step, so exclude `False`, `not_initialized`, `disabled`, and maybe `idle` (I have to double check that one).

I'll push the last fixes so we can more forward.

File: addons/mail_bot/models/ir_http.py [hunk 13]
```diff
@@ -1,14 +0,0 @@
...
-
-
-class IrHttp(models.AbstractModel):
-    _inherit = ['ir.http']
-
-    def session_info(self):
-        res = super().session_info()
-        if self.env.user._is_internal():
-            res['odoobot_initialized'] = self.env.user.odoobot_state not in [False, 'not_initialized']
-        return res
```
Comment (seb-odoo) [inline addons/mail_bot/models/ir_http.py:14]: Note: cleanup, this was dead code

File: addons/mail_bot/models/mail_thread.py [hunk 14]
```diff
@@ -1,12 +0,0 @@
...
-
-from odoo import models
-
-
-class MailThread(models.AbstractModel):
-    _inherit = ['mail.thread']
-
-    def _message_post_after_hook(self, message, msg_vals):
-        self.env['mail.bot']._apply_logic(self, msg_vals)
-        return super(MailThread, self)._message_post_after_hook(message, msg_vals)
```
Comment (seb-odoo) [inline addons/mail_bot/models/mail_thread.py:12]: Note: cleanup, this had nothing to do on a generic thread override. Moved to channel.

File: addons/mail_bot/models/mail_bot.py [hunk 15]
```diff
@@ -288,16 +292,6 @@ def _body_contains_emoji(self, body):
...
 
-    def _is_bot_pinged(self, values):
-        odoobot_id = self.env['ir.model.data']._xmlid_to_res_id("base.partner_root")
-        return odoobot_id in values.get('partner_ids', [])
-
-    def _is_bot_in_private_channel(self, record):
-        odoobot_id = self.env['ir.model.data']._xmlid_to_res_id("base.partner_root")
-        if record._name == 'discuss.channel' and record.channel_type == 'chat':
-            return odoobot_id in record.with_context(active_test=False).channel_partner_ids.ids
-        return False
```
Comment (seb-odoo) [inline addons/mail_bot/models/mail_bot.py:299]: Note: cleanup, used only once or written in much more complex ways than necessary.

File: addons/mail_bot/models/mail_bot.py [hunk 16]
```diff
@@ -11,26 +11,28 @@
...
+         :param channel: the discuss channel where the user message was posted/odoobot will answer.
          :param values: msg_values of the message_post or other values needed by logic
          :param command: the name of the called command if the logic is not triggered by a message_post
         """
+        channel.ensure_one()
         odoobot_id = self.env['ir.model.data']._xmlid_to_res_id("base.partner_root")
-        if len(record) != 1 or values.get("author_id") == odoobot_id or values.get("message_type") != "comment" and not command:
+        if values.get("author_id") == odoobot_id or values.get("message_type") != "comment" and not command:
             return
-        if self._is_bot_pinged(values) or self._is_bot_in_private_channel(record):
```
Comment (seb-odoo) [inline addons/mail_bot/models/mail_bot.py:27]: Note: cleanup, condition `_is_bot_pinged` irrelevant as `_is_bot_in_private_channel` was always checked in `_get_answer`.

File: addons/mail/static/src/core/common/store_service.js [hunk 17]
```diff
@@ -174,18 +172,9 @@ export class Store extends BaseStore {
...
              * In each group, thread with most recent message comes first
              */
-            const aOdooBot = a.isCorrespondentOdooBot;
-            const bOdooBot = b.isCorrespondentOdooBot;
-            if (aOdooBot && !bOdooBot) {
-                return 1;
-            }
-            if (bOdooBot && !aOdooBot) {
-                return -1;
-            }
```
Comment (seb-odoo) [inline addons/mail/static/src/core/common/store_service.js:188]: Note: this was just not a good idea, removed all odoobot execeptions.

File: addons/mail/static/src/core/public_web/store_service_patch.js [hunk 18]
```diff
@@ -14,13 +14,8 @@ patch(Store.prototype, {
         this.discuss = { activeTab: "main" };
         this.env.bus.addEventListener(
             "discuss.channel/new_message",
-            ({ detail: { channel, message } }) => {
-                if (this.env.services.ui.isSmall || message.isSelfAuthored) {
-                    return;
-                }
-                if (channel.isCorrespondentOdooBot && this.odoobotOnboarding) {
-                    // this cancels odoobot onboarding auto-opening of chat window
-                    this.odoobotOnboarding = false;
```
Comment (seb-odoo) [inline addons/mail/static/src/core/public_web/store_service_patch.js:23]: Note: this was just not a good idea, removed all odoobot execeptions.
Added `silent` which is only enabled for onboarding specifically.
Other potentially important messages from OdooBot are no longer ignored.

File: addons/mail/static/src/core/public_web/thread_model_patch.js [hunk 19]
```diff
@@ -36,9 +36,6 @@ patch(Thread.prototype, {
      * @param {import("models").Message} message
      */
     notifyMessageToUser(message) {
-        if (this.isCorrespondentOdooBot) {
-            return;
-        }
```
Comment (seb-odoo) [inline addons/mail/static/src/core/public_web/thread_model_patch.js:41]: Note: this was just way too drastic and prevented any notif from OdooBot, even in other potentially important flows.
It completely shadowed condition in `discuss.channel/new_message` too.

We might need to backport this part actually.

**Review by alexkuhn — APPROVED** [review]
@robodoo r+
